src/main/scala/com/databricks/labs/overwatch/env/Database.scala

```diff
@@ -232,6 +232,99 @@ class Database(config: Config) extends SparkSessionWrapper {
     }
   }
 
+  /**
+   * Get the merge condition based on the maxMergeScanDates on target's immutable columns
+   * @param df
+   * @param target
+   * @param maxMergeScanDates
+   * @return
+   */
+  def getMergeCondition(df: DataFrame, target: PipelineTable, maxMergeScanDates: Array[String] = Array()): String = {
+    val immutableColumns = (target.keys ++ target.incrementalColumns).distinct
+
+    val datePartitionFields = getPartitionedDateField(df, target.partitionBy)
+    val explicitDatePartitionCondition = if (datePartitionFields.nonEmpty & maxMergeScanDates.nonEmpty) {
+      s" AND target.${datePartitionFields.get} in (${maxMergeScanDates.mkString("'", "', '", "'")})"
+    } else ""
+    val mergeCondition: String = immutableColumns.map(k => s"updates.$k = target.$k").mkString(" AND ") + " " +
+      s"AND target.organization_id = '${config.organizationId}'" + // force partition filter for concurrent merge
+      explicitDatePartitionCondition // force right side scan to only scan relevant dates
+
+    mergeCondition
+  }
+
+  /**
+   * Function to write the batch data in delta format
+   * @param df
+   * @param target
+   * @param maxMergeScanDates
+   */
+  def batchWriter(df: DataFrame, target: PipelineTable, maxMergeScanDates: Array[String] = Array()): Unit = { // DELTA MERGE / UPSERT
+    val deltaTarget = DeltaTable.forPath(spark,target.tableLocation).alias("target")
+    val updatesDF = df.alias("updates")
+    val mergeCondition = getMergeCondition(df, target, maxMergeScanDates)
+
+    val mergeDetailMsg =
+      s"""
+         |Beginning upsert to ${target.tableFullName}.
+         |MERGE CONDITION: $mergeCondition
+         |""".stripMargin
+    logger.log(Level.INFO, mergeDetailMsg)
+    spark.conf.set("spark.databricks.delta.commitInfo.userMetadata", config.runID)
+    // TODO -- when DBR 9.1 LTS GA, use LSM (low-shuffle-merge) to improve pipeline
+    deriveDeltaMergeBuilder(deltaTarget, updatesDF, mergeCondition, target)
+      .execute()
+
+    spark.conf.unset("spark.databricks.delta.commitInfo.userMetadata")
+  }
+
+  /**
+   * Function to write the streaming data
+   * @param df
+   * @param target
+   */
+  def streamWriter(df: DataFrame, target: PipelineTable): Unit = {
```

Comment:
Since this function does both stream and batchWrites, I'd switch it to dataframeWriter or something else. And change batchWriter to mergeWriter or something else

Reply:
done